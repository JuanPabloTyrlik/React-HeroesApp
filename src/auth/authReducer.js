import { TYPES } from '../types/types';

const initialState = {
    logged: false,
};

export const authReducer = (state = initialState, { type, payload }) => {
    switch (type) {
        case TYPES.LOGIN:
            return { ...payload, logged: true };
        case TYPES.LOGOUT:
            return { logged: false };
        default:
            return state;
    }
};
