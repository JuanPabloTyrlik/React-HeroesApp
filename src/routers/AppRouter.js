import React from 'react';
import { useContext } from 'react';
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom';
import { AuthContext } from '../auth/AuthContext';
import { LoginScreen } from '../components/login/LoginScreen';
import { DashboardRouter } from './DashboardRouter';
import { PrivateRoute } from './PrivateRoute';

export const AppRouter = () => {
    const {
        user: { logged },
    } = useContext(AuthContext);

    return (
        <Router>
            <div>
                <Switch>
                    <Route exact path="/login" component={LoginScreen} />
                    <PrivateRoute
                        path="/"
                        component={DashboardRouter}
                        isAuthenticated={logged}
                    />
                </Switch>
            </div>
        </Router>
    );
};
