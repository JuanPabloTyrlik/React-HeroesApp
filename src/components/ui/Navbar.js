import React from 'react';
import { useContext } from 'react';
import { Link, NavLink, useHistory } from 'react-router-dom';
import { AuthContext } from '../../auth/AuthContext';
import { SearchContext } from '../../routers/DashboardRouter';
import { TYPES } from '../../types/types';

export const Navbar = () => {
    const { searchValue, setSearchValue } = useContext(SearchContext);
    const { user, dispatch } = useContext(AuthContext);
    const history = useHistory();

    const handleChange = (e) => {
        e.preventDefault();
        setSearchValue(e.target.value);
    };

    const handleReset = () => {
        setSearchValue('');
    };

    const handleLogOut = () => {
        dispatch({ type: TYPES.LOGOUT });
        history.replace('/login');
    };

    return (
        <nav className="navbar navbar-expand-lg navbar-dark bg-dark">
            <Link className="navbar-brand" to="/" onClick={handleReset}>
                Comics
            </Link>

            <div className="navbar-collapse">
                <div className="navbar-nav">
                    <NavLink
                        activeClassName="active"
                        className="nav-item nav-link"
                        exact
                        to="/marvel"
                        onClick={handleReset}
                    >
                        Marvel
                    </NavLink>

                    <NavLink
                        activeClassName="active"
                        className="nav-item nav-link"
                        exact
                        to="/dc"
                        onClick={handleReset}
                    >
                        DC
                    </NavLink>
                </div>
            </div>

            <div className="navbar-collapse collapse w-100 order-3 dual-collapse2">
                <ul className="navbar-nav ml-auto">
                    <form
                        className="form-inline my-2 my-lg-0"
                        onSubmit={(e) => e.preventDefault()}
                    >
                        <input
                            className="form-control mr-sm-2"
                            type="search"
                            placeholder="Search Hero"
                            aria-label="Search"
                            onChange={handleChange}
                            value={searchValue}
                        />
                    </form>
                    <span className="navbar-text text-success">
                        {user.name}
                    </span>
                    <button
                        className="nav-item nav-link btn"
                        onClick={handleLogOut}
                    >
                        Logout
                    </button>
                </ul>
            </div>
        </nav>
    );
};
