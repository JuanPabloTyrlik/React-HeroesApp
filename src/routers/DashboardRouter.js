import React, { useState } from 'react';
import { createContext } from 'react';
import { Redirect, Route, Switch } from 'react-router-dom';
import { DcScreen } from '../components/dc/DcScreen';
import { HeroScreen } from '../components/heroes/HeroScreen';
import { MarvelScreen } from '../components/marvel/MarvelScreen';
import { SearchScreen } from '../components/serach/SearchScreen';
import { Navbar } from '../components/ui/Navbar';

export const SearchContext = createContext();

export const DashboardRouter = () => {
    const [searchValue, setSearchValue] = useState('');

    return (
        <>
            <SearchContext.Provider value={{ searchValue, setSearchValue }}>
                <Navbar />
                <div className="container mt-2">
                    <Switch>
                        <Route exact path="/marvel" component={MarvelScreen} />
                        <Route exact path="/dc" component={DcScreen} />
                        <Route exact path="/search" component={SearchScreen} />
                        <Route
                            exact
                            path="/hero/:heroId"
                            component={HeroScreen}
                        />
                        <Redirect to="/marvel" />
                    </Switch>
                </div>
            </SearchContext.Provider>
        </>
    );
};
