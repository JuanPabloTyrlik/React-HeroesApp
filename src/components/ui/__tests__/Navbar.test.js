import React from 'react';
import { mount } from 'enzyme';
import { Router } from 'react-router-dom';
import { Navbar } from '../Navbar';
import { AuthContext } from '../../../auth/AuthContext';
import { SearchContext } from '../../../routers/DashboardRouter';
import { TYPES } from '../../../types/types';
import { createMemoryHistory } from 'history';

describe('Tests on Navbar', () => {
    const searchContextValue = {
        searchValue: '',
        setSearchValue: jest.fn(),
    };

    const authContextValue = {
        user: {
            logged: true,
            name: 'Juan Pablo',
        },
        dispatch: jest.fn(),
    };

    const history = createMemoryHistory();
    history.push('/marvel');
    history.replace = jest.fn();

    const wrapper = mount(
        <AuthContext.Provider value={authContextValue}>
            <SearchContext.Provider value={searchContextValue}>
                <Router history={history}>
                    <Navbar />
                </Router>
            </SearchContext.Provider>
        </AuthContext.Provider>
    );

    beforeAll(() => jest.clearAllMocks());

    test('should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });
    test("should display logged user's name", () => {
        expect(wrapper.find('span.navbar-text.text-success').text()).toMatch(
            authContextValue.user.name
        );
    });
    test('should logout user', () => {
        wrapper.find('button').simulate('click');
        expect(authContextValue.dispatch).toHaveBeenCalledWith({
            type: TYPES.LOGOUT,
        });
        expect(history.replace).toHaveBeenCalledWith('/login');
    });
});
