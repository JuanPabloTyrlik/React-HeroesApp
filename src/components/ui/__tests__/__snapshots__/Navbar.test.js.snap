// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on Navbar should render correctly 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-dark bg-dark"
>
  <a
    className="navbar-brand"
    href="/"
    onClick={[Function]}
  >
    Comics
  </a>
  <div
    className="navbar-collapse"
  >
    <div
      className="navbar-nav"
    >
      <a
        aria-current="page"
        className="nav-item nav-link active"
        href="/marvel"
        onClick={[Function]}
        style={Object {}}
      >
        Marvel
      </a>
      <a
        aria-current={null}
        className="nav-item nav-link"
        href="/dc"
        onClick={[Function]}
      >
        DC
      </a>
      <a
        aria-current={null}
        className="nav-item nav-link"
        href="/search"
        onClick={[Function]}
      >
        Search
      </a>
    </div>
  </div>
  <div
    className="navbar-collapse collapse w-100 order-3 dual-collapse2"
  >
    <ul
      className="navbar-nav ml-auto"
    >
      <form
        className="form-inline my-2 my-lg-0"
        onSubmit={[Function]}
      >
        <input
          aria-label="Search"
          className="form-control mr-sm-2"
          onChange={[Function]}
          placeholder="Search Hero"
          type="search"
          value=""
        />
      </form>
      <span
        className="navbar-text text-success"
      >
        Juan Pablo
      </span>
      <button
        className="nav-item nav-link btn"
        onClick={[Function]}
      >
        Logout
      </button>
    </ul>
  </div>
</nav>
`;
